lang/funcs/collection.go

```diff
@@ -695,8 +701,10 @@ var LookupFunc = function.New(&function.Spec{
 					return cty.NumberVal(v.AsBigFloat()), nil
 				case ty.Equals(cty.Bool):
 					return cty.BoolVal(v.True()), nil
+				case ty.IsObjectType() || ty.IsListType() || ty.IsMapType():
+					return v, nil
 				default:
-					return cty.NilVal, errors.New("lookup() can only be used with maps of primitive types")
+					return cty.NilVal, fmt.Errorf("lookup() cannot be used with type %s", ty.FriendlyName())
```

Comment:
This `switch` statement seems to ultimately just be a bunch of no-ops: if it's a string then convert it to a string, if it's a number convert it to a number, etc.

I'm wondering if we could just replace this whole thing with `return mapVar.Index(cty.StringVal(lookupKey)), nil` and thus be completely type agnostic, just like the language's built-in index operator. Do you remember what we put this here for originally? Perhaps it was just inherited from the 0.11 implementation of `lookup`? :thinking:

Reply:
That's beautiful! And no, I have no recollection of why this was done this way. I implemented your suggestion and the tests continue to pass, so hooray for simplification!